src/shared/views/BuildingFootprint/Validator.js

```diff
@@ -2,145 +2,300 @@
-        if (done) {
-            // no more tasks in the group, commit results and go back to menu
-            commitCompletedGroup();
+        const { currentTaskIndex } = this.state;
+        if (result !== null && result !== undefined) {
+            // the user tapped a button, save result and update how far
+            // we've worked
+            submitResult(result, this.expandedTasks[currentTaskIndex].taskId);
+            if (currentTaskIndex > this.tasksDone) {
+                this.tasksDone = currentTaskIndex;
+            }
+        } else if (currentTaskIndex > this.tasksDone) {
+            // no result provided, the user just tried to swipe forward
+            // past the last task completed, just ignore this swipe
+            // but return true to indicate we can't go any further
+            return true;
+        }
+        if (currentTaskIndex + 1 >= this.expandedTasks.length) {
+            // no more tasks in the group, show the "LoadMore" screen
+            completeGroup();
+            return false;
         }
-        this.tasksDone += 1;
-        updateProgress(this.tasksDone / group.numberOfTasks);
-        this.setState({ currentTaskId: value });
+        updateProgress(currentTaskIndex / group.numberOfTasks);
+        this.setState({ currentTaskIndex: currentTaskIndex + 1 });
+        return false;
     };
 
-    // eslint-disable-next-line class-methods-use-this
-    *makeNextTaskGenerator(
-        tasks: Array<BuildingFootprintTaskType>,
-    ): taskGenType {
-        // generator function that picks the next task to work on
-        // we cannot assume any specific order of taskId in the group
-        const taskIds = tasks.map((t) => t.taskId);
-        let i;
-        // eslint-disable-next-line no-plusplus
-        for (i = 0; i < taskIds.length; i++) {
-            yield taskIds[i];
+    previousTask = (): boolean => {
+        // update state to point to the previous task and return a
+        // bool indicating whether we've reached the end of the array of tasks
+        const { group, updateProgress } = this.props;
+        const { currentTaskIndex } = this.state;
+        if (currentTaskIndex > 0) {
+            updateProgress(currentTaskIndex / group.numberOfTasks);
+            this.setState({ currentTaskIndex: currentTaskIndex - 1 });
+            return false;
         }
-    }
+        return true;
+    };
 
+    /* eslint-disable global-require */
     render = () => {
-        const { group, project } = this.props;
-        const { currentTaskId } = this.state;
-        if (!group.tasks) {
+        const { project, results } = this.props;
+        const { currentTaskIndex } = this.state;
+        if (!this.expandedTasks) {
             return <LoadingIcon />;
         }
-        const currentTask = group.tasks.find((t) => t.taskId === currentTaskId);
+        const currentTask = this.expandedTasks[currentTaskIndex];
+        // if tasks have a center attribute, we know they're grouped by 9
```

Comment:
maybe for clarification: for which tasks do we expect a center attribute? only for the google ones?

Reply:
Yes, for now, we only use it with google imagery, if it exists (it's not required, it's just possible to add it for optimisation). I've added a bit of explanation in the code, hopefully it helps us remember what we did here :)